bootloader/lkenv/lkenv_test.go

```diff
@@ -0,0 +1,183 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2014-2015 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package lkenv_test
+
+import (
+	"io/ioutil"
+	"path/filepath"
+	"testing"
+
+	. "gopkg.in/check.v1"
+
+	"github.com/snapcore/snapd/bootloader/lkenv"
+)
+
+// Hook up check.v1 into the "go test" runner
+func Test(t *testing.T) { TestingT(t) }
+
+type lkenvTestSuite struct {
+	envPath    string
+	envPathbak string
+}
+
+var _ = Suite(&lkenvTestSuite{})
+
+func (l *lkenvTestSuite) SetUpTest(c *C) {
+	l.envPath = filepath.Join(c.MkDir(), "snapbootsel.bin")
+	l.envPathbak = l.envPath + "bak"
+}
+
+func (l *lkenvTestSuite) TestSet(c *C) {
+	env := lkenv.NewEnv(l.envPath)
+	c.Check(env, NotNil)
+
+	env.Set("snap_mode", "try")
+	c.Check(env.Get("snap_mode"), Equals, "try")
+}
+
+func (l *lkenvTestSuite) TestSave(c *C) {
+	buf := make([]byte, 4096)
+	err := ioutil.WriteFile(l.envPathbak, buf, 0644)
+	c.Assert(err, IsNil)
+	l.TestSaveNoBak(c)
+}
+
+func (l *lkenvTestSuite) TestSaveNoBak(c *C) {
+	buf := make([]byte, 4096)
+	err := ioutil.WriteFile(l.envPath, buf, 0644)
+	c.Assert(err, IsNil)
+
+	env := lkenv.NewEnv(l.envPath)
+	c.Check(env, NotNil)
+
+	env.Set("snap_mode", "trying")
+	env.Set("snap_kernel", "kernel-1")
+	env.Set("snap_try_kernel", "kernel-2")
+	env.Set("snap_core", "core-1")
+	env.Set("snap_try_core", "core-2")
+	env.Set("snap_gadget", "gadget-1")
+	env.Set("snap_try_gadget", "gadget-2")
+
+	err = env.Save()
+	c.Assert(err, IsNil)
+
+	env2 := lkenv.NewEnv(l.envPath)
```

Comment:
I think in addition to this test we want something that stores the expected bytes data in some variable and compares the written structure byte-to-byte. This test here is nice but if e.g. alignment breaks we won't notice because the same code that reads and writes is used in the test. To fix this I think we need an extra test against some blob (in the form of  ` var  expectedStructData = []byte{0x11,0x22,...}` (unless that is too long in which case we could use an external file).

Reply:
@mvo5 I added now gzipped sample data test case.
Let me know if more tests are needed :)